### PR TITLE
Assume `build.rs` in the same directory as `Cargo.toml` is a build script (unless explicitly told not to)

### PR DESCRIPTION
**Pull request by pwoolcoc**
So, in May I posted a question in the #cargo IRC room: https://botbot.me/mozilla/cargo/2016-05-26/?msg=66770068&page=1

This PR does what was discussed there. If `cargo` sees a `build.rs` in the same directory as the `Cargo.toml`, it will assume `build.rs` is a build script unless `package.build = false`. Just for completeness I also made `build = true` mean the same as `build = "build.rs"` but I'm not sure if that is okay or not.

### CODE REVIEW

**Comment by steveklabnik** [comment]
yessssssss i asked @alexcrichton about this a while back and he was 👍 at the time. I think this is great.

File: src/cargo/util/toml.rs [hunk 1]
```diff
@@ -540,7 +546,11 @@ impl TomlManifest {
         }
 
         // processing the custom build script
-        let new_build = project.build.as_ref().map(PathBuf::from);
+        let manifest_file = util::important_paths::find_root_manifest_for_wd(None, &layout.root)
```
Comment (pwoolcoc) [inline src/cargo/util/toml.rs]: Looking this over again, I'm wondering if I actually have to do this, or if I can just assume `base_dir == layout.root`?

**Comment by pwoolcoc** [comment]
@steveklabnik :) everyone who commented in #cargo seemed to support it, but it was a while ago, so I'm hoping the support is still there.

**Comment by alexcrichton** [comment]
Thanks for the PR! The implementation looks great to me as well.

My only hesitation here is that we risk breaking projects when this lands. Projects can fix builds with `build = false` in `Cargo.toml`, but I think that'll end up getting rejected on the stable/beta channels. In that sense, if a crate has a `build.rs` but it's not a build script, then you can't keep that layout and continue to work on stable/beta/nightly.

I'd also want to think through the backcompat story here with all crates published on crates.io. It may be worth doing a survey there to see how many crates have a `build.rs` file which *isn't* a build script, if any. If that number is 0, then our lives are much easier! If that number is greater, we may want to proactively reach out to authors to head off the breakage.

@pwoolcoc would you be willing to scrape crates.io and test for `build.rs` scripts in roots which aren't build scripts?

**Comment by pwoolcoc** [comment]
@alexcrichton sure, I can do that.

Would it preferable to change it slightly to only print a deprecation warning when there is a build.rs that isn't a build script, and land the rest of the change after a couple releases?

**Comment by steveklabnik** [comment]
I haven't read the patch, but to be clear, the way in which this would
break is if someone had specially configured a `build.rs` at the top level
to be source, correct? I'd imagine this would be very, very tiny, given
almost all source files are in `src`, generally speaking.

**Comment by pwoolcoc** [comment]
@steveklabnik yea, if someone has a crate with a `build.rs` file in the same directory as `Cargo.toml`, and the `build.rs` is _not_ a build script, then this patch will try to treat it as a build script. The user would have to add `build = false` to the `package` section of their `Cargo.toml` in order to prevent this.

**Comment by alexcrichton** [comment]
@pwoolcoc 

> Would it preferable to change it slightly to only print a deprecation warning when there is a build.rs that isn't a build script, and land the rest of the change after a couple releases?

If we have no existing `build.rs` scripts that *aren't* build scripts, I think we can avoid this. Otherwise yeah I think we'll need to take this route. I figure we should just get a handle on the situation first so we can categorize our options.

@steveklabnik 

> the way in which this would break is if someone had specially configured a `build.rs` at the top level to be source, correct?

Indeed! I agree that it should be rare if ever, but I just want to make sure!

**Comment by pwoolcoc** [comment]
@alexcrichton ok, I'm finishing up the few crates that my script couldn't automatically check, but here are the preliminary results. There are 12 crates that have a `build.rs` in the same directory as a `Cargo.toml`, without an accompanying `build = "build.rs"` in the `Cargo.toml` file. However, _all_ of the `build.rs` are build scripts that are just not being used. None of them are actually source code for the project. However, this still means that their builds might break from this change.

I have about ~10 crates that I need to check manually, I'll update here if necessary.

The 12 crates that I mentioned above are:

>pcb				https://github.com/ubsan/pcb/pcb-c

>libsqlite3-sys			https://github.com/jgallagher/rusqlite

>sphinxad-sys			https://github.com/TrionProg/sphinxad-sys

>webkitten			https://github.com/kattrali/webkitten

>core-nightly                    https://github.com/rust-lang/rust

>al-sys			        https://github.com/jpernst/alto/

>roboime-next			https://github.com/roboime/roboime-next/

>hoodlum				https://github.com/tcr/hoodlum

>static-server			https://github.com/DenisKolodin/static-server

>strcursor			https://github.com/DanielKeep/strcursor

>rustproof			https://github.com/Rust-Proof/rustproof 

>openal-rs			https://github.com/jpernst/alto

**Comment by alexcrichton** [comment]
@pwoolcoc thanks for the investigation! Looking at those thought I think that this change would break most of those crates unfortunately. Most of the build scripts have an `extern crate` directive which isn't listed in `[build-dependencies]`.

Given that we may need to pursue a strategy like:

* Add knowledge of `build = <bool>`
* If `build` isn't specified and `build.rs` exists, generate a warning

We probably need to let that bake for quite some time, and then after that's in we can enable the auto-detection. What do you think?

**Comment by pwoolcoc** [comment]
@alexcrichton sounds good, I suspected as much, so I have a patch ready to go for this. So to be clear, the new behavior would be:

  1. If `build = "bulid.rs"` is present, build script continues to work as normal, regardless of the existence of `build.rs`
  2. if `build = true` is present, it is equivalent to `build = "build.rs"`, regardless of the existence of `build.rs`
  3. If `bulid = false` is present, it is equivalent to not having a `build` value at all, regardless of the existence of `build.rs`
  4. If `build = ...` is not present, and a `build.rs` file is present, a warning would be printed.

Does that sounds right? I'll push my last commit so you can look it over, too.

**Comment by alexcrichton** [comment]
@pwoolcoc sounds exactly right to me!

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by pwoolcoc** [comment]
@alexcrichton sorry to revive the old thread, just had a question about this. What is the procedure for going back to finish features like this a few releases down the line? Should I open a tracking issue for it? Or just put a reminder in my calendar for March and ping you again after a couple release cycles?

**Comment by alexcrichton** [comment]
@pwoolcoc oh we unfortunately don't have a super principled way of handling this in Cargo. For now though want to open a tracking issue here?

**Pull request merged**